Add hardware-metrics to semantic conventions

bertysentry opened the pull request:
## Changes

Added hardware metrics to semantic conventions

Related to issue #2340

tigrannajaryan commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,340 @@
+
+## Metric Instruments
+
+### `hw.host.` - Physical host metrics
```
How is the host identified? Do we expect that these metrics are accompanied by a `Resource` with `host.*` attributes?

tigrannajaryan replied:
Nevermind, I see it is already answered in the PR.

tigrannajaryan commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,340 @@
+
+| Name                          | Description                                                              | Units                                  | Instrument Type ([*](README.md#instrument-types)) | Value Type | Attribute Key(s) | Attribute Values |
+| ----------------------------- | ------------------------------------------------------------------------ | -------------------------------------- | ------------------------------------------------- | ---------- | ---------------- | ---------------- |
+| `hw.host.ambient_temperature` | Ambient (external) temperature of the physical host                      | °C                                     | Gauge                                             | Double     |                  |                  |
```
>°C

Please use [UCUM c/s](https://ucum.org/ucum.html#si) unit specifiers instead of Print unit specifier (e.g. `Cel`).

bertysentry replied:
Oops, will do!

tigrannajaryan reviewed (commented):
@bertysentry Thank you for the PR.

The linked issue has not been discussed yet. I think we likely want to have hardware metrics in the spec, but I would want to hear from other spec approvers if they agree that it is a good idea to have semantic conventions for the hardware metrics.

If the community feels these do belong to the spec then we will need a detailed review which may take some time due to the PR's size. Unless there is a reason that all of these must be merged together it may be easier to make progress if this is presented as a few separate PRs, one per hardware area (once there is a general agreement that we do want hardware metrics in the spec).

A few generic questions/comments:
- Please rebase from main and fix the build.
- Do we want `hw.` as a prefix and not `system.`? The linked issue talks about adding e.g. `system.sensor.temp`. Also, do we want abbreviated `hw.` or `hardware.`? Any arguments in favour of against any of these approaches?
- How does this proposal compare with any existing industry standards for hardware metrics reporting (if any exist)?

tigrannajaryan commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,340 @@
+| `hw.host.energy`              | Total energy consumed by the physical host, in joules                    | J                                      | Counter                                           | Int64      |                  |                  |
+| `hw.host.heating_margin`      | Distance in degrees Celsius to the closest temperature warning threshold | °C                                     | Gauge                                             | Double     |                  |                  |
+| `hw.host.power`               | Power consumed by the physical host in Watts                             | W                                      | Gauge                                             | Double     |                  |                  |
+| `hw.host.status`              | Physical host overall operational status                                 | { 0 = OK ; 1 = Degraded ; 2 = Failed } | Gauge                                             | Int        |                  |                  |
```
>{ 0 = OK ; 1 = Degraded ; 2 = Failed }

This does not look like a Unit. I assume this is intended to reflect the state, which is one of the 3 possible values, so perhaps put this in the Value Type column.

Alternatively, could we have an attribute that describes the state and the value to show a metric for that state? (This can be analogous to `up` metrics which is either 0 or 1).

bertysentry replied:
It is not a standard unit, you're right. I tried to follow the other semantic conventions in OpenTelemetry, where non-standard units are surrounded with curly brackets (e.g. `{packets}`). I'd argue it's better to describe the possible values in the unit itself so it ends up in the UI (say on the Y-axis of a graph) and the user can interpret the numeric values.

[OpenMetrics recommends to map enumerations to integer values](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#gauge) rather than attributes for better efficiency. Users seem to prefer that we do not to require the collector to push 3 *DataPoints* (1, 0, and 0) for each OK status value.

Also, see [my comment on this issue](https://github.com/open-telemetry/opentelemetry-specification/issues/1711#issuecomment-1114303004) 😉 If we decided of such a convention, it would allow UI developers to automatically interpret such enum values.

pirgeo replied:
I would argue against using the Unit to do this. In my mind, the Unit should be something that is just read by the UI, and if the UI knows how to display the Unit, it should do so. Having the UI parse such a string and then generate the respective code to show the right thing seems very fragile to me. Furthermore, backends that parse units and do something with them will have to work around the new format. This might be okay if you are developing your backend yourself, but in that case you can use every field as you see fit anyway. I don't think this is a good fit for a standard published by OpenTelemetry, but maybe I am misunderstanding what the unit field was intended to do in the first place, so please correct me if I am wrong. If so, I think we should add a standard format to the spec.

Additionally, using unit in this way would send the signal that it is generally okay (maybe even endorsed by OpenTelemetry) to use the Unit to represent Enums. While your current proposal fits the 63 character limit for unit, other translations from enum to unit might not. The users would then be forced to work around it by adding their values as dimensions anyway, and I don't think we want to endorse such inconsistencies in OpenTelemetry, if we can help it.
Finally, changing the unit at a later time (e.g. adding a new enum value) might need to change the code your UI uses to display charts. Adding a new dimension value on the contrary is a well understood change in the data stream, and is preferrable in my opinion.

Also, maybe some compression could help with the downside of having to report 3 data points instead of one. Potentially, something like the [column-based OTLP format](https://github.com/open-telemetry/oteps/pull/171) could be helpful.

All of these considerations lead me to the conclusion that adding the state as a dimension value is still the preferred method. I don't think Unit is the right place to do this.

bertysentry replied:
Good point about the 63-char limit. Also, you're right: this is just a workaround waiting for a true *Enum* type in OpenTelemetry metrics. See [my comment on the corresponding issue to create an *Enum* metric value type](https://github.com/open-telemetry/opentelemetry-specification/issues/1711#issuecomment-1117349513).

> In the meantime, assuming OpenTelemetry adds an *Enum* value type in the future, I would recommend that developers use a single *Gauge int* metric to represent mutually exclusive enums, as it will be much easier to convert later a single *int* metric to an *Enum*.

Conclusion: the *status* metric can remain as a *Gauge int* until we get *Enum*, but the `{ 0 = OK ; 1 = Degraded ; 2 = Failed }` unit should be left out of the semantic convention, as it's not an OpenTelemetry recommendation.

tigrannajaryan replied:
I agree that it is very valuable to have the enum statuses in the metrics. My only concern is that we will add the enums in a way that we don't like and then we will break it very soon when we come up with a proper way to represent enums.

reyang commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,340 @@
+| Attribute Key | Description                                                                                                   | Example                             |
+| ------------- | ------------------------------------------------------------------------------------------------------------- | ----------------------------------- |
+| `id`          | A unique identifier for the hardware component in the monitored host                                          | `win32battery_battery_testsysa33_1` |
+| `label`       | An easily-recognizable name for the hardware component                                                        | `eth0`                              |
```
Would `name` be a better name? `label` is widely used by Prometheus https://prometheus.io/docs/practices/naming/, which could introduce unnecessary confusion.

bertysentry replied:
True, it is probably better.

reyang commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,340 @@
+| Name                          | Description                                                              | Units                                  | Instrument Type ([*](README.md#instrument-types)) | Value Type | Attribute Key(s) | Attribute Values |
+| ----------------------------- | ------------------------------------------------------------------------ | -------------------------------------- | ------------------------------------------------- | ---------- | ---------------- | ---------------- |
+| `hw.host.ambient_temperature` | Ambient (external) temperature of the physical host                      | °C                                     | Gauge                                             | Double     |                  |                  |
+| `hw.host.energy`              | Total energy consumed by the physical host, in joules                    | J                                      | Counter                                           | Int64      |                  |                  |
```
Would we want Int64 or Double? Asking this because there could be low power consumption hosts.

bertysentry replied:
Good point! 

We need the full precision of *Int64* because we will need to be able to calculate precise difference between very high values of the counter to calculate the power in Watts: `delta(hw.host.energy) / delta(time)`.

reyang commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,340 @@
+| `hw.memory.predicted_failure` | Whether the memory module is predicting an imminent failure | { 0 = false ; 1 = true }               | Gauge                                             | Int        |               |                  |
+| `hw.memory.status`            | Memory module operational status                            | { 0 = OK ; 1 = Degraded ; 2 = Failed } | Gauge                                             | Int        |               |                  |
+
+Additional attributes:
```
Would you consider additional attributes such as memory banks / affinity group in a [NUMA](https://en.wikipedia.org/wiki/Non-uniform_memory_access) environment?

bertysentry replied:
I guess any extra attributes okay, but they are not required, like timings, etc. We'd recommend to put the bank number (or location on the mainboard) in the `name` attribute.

reyang commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,340 @@
+
+**Description:** A physical network interface, or a network interface controller
+(NIC), excluding virtual adapters and loopbacks. Examples: an Ethernet adapter,
+an HBA, an fiber channel port or a Wifi adapter.
```
Minor:

```suggestion
an HBA, an fiber channel port or a Wi-Fi adapter.
```

reyang commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,340 @@
+| Attribute Key      | Description                                                | Example             |
+| ------------------ | ---------------------------------------------------------- | ------------------- |
+| `model`            | Descriptive model name                                     |                     |
+| `logical_address`  | Logical address of the adapter (e.g. IP address, or WWPN)  | `172.16.8.21`       |
```
Consider `logical_addresses` - a single NIC could have multiple IPs.

bertysentry replied:
True!

reyang commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,340 @@
+
+### `hw.physical_disk.`- Physical disk metrics
+
+**Description:** Physical hard drive (HDD or SDD)
```
Consider borrowing from [S.M.A.R.T.](https://en.wikipedia.org/wiki/S.M.A.R.T.)?

bertysentry replied:
Currently, only the `hw.physical_disk.predicted_failure` metric would report the overall S.M.A.R.T. status (as an aggregated status of all available metrics on the drive). But I take your point: we could add a generic `hw.physical_disk.smart` metric with a `smart_attribute` attribute.

reyang reviewed (approved):
Major suggestions:
1. Use standard units.
2. Consider/Explore putting metadata as Resource instead of Attributes (most of the data won't change - unless the hardware changed).
3. Reference to other standards (e.g. S.M.A.R.T.).

I've approved the PR because all of the above suggestions can be addressed in follow up PRs. The overall direction looks good to me.

bertysentry commented:
Thank you for the time you spent on this review and your comments, @tigrannajaryan

Let me answer your questions:

> Do we want hw. as a prefix and not system.? The linked issue talks about adding e.g. system.sensor.temp. 

Having something separate from `system.` is important because `system` represents the operating system and its performance metrics, which probably runs in a virtual machine while `hw` represents a physical host only. 

Example: at the hardware level, you don't consider CPU time utilization (as it's purely an OS metric). So `system.cpu.*` and `hw.cpu.` are 2 very different beasts. `system.cpu` represents a CPU core, while `hw.cpu` represents the "real" CPU package (which includes several cores). 

If we mix `system` and `hw`, we will be mixing things that are different in nature and the metrics will be difficult to leverage in dashboards. 

Other examples are disks (the physical disk behind a RAID controller are different from the disks seen by the operating system) and network interfaces (the OS has loopback and plenty of virtual adapters for aggregation, VPNs, etc.).

> Do we want abbreviated hw. or hardware.? Any arguments in favour of against any of these approaches?

Currently, we have `system` (full word) and `db` (abbreviation), so both are acceptable. I tend to prefer `hw` because it's shorter while still perfectly understandable. Since it's shorter, it consumes less energy and thus emits less carbon 😊

> How does this proposal compare with any existing industry standards for hardware metrics reporting (if any exist)?

My company has been developing and maintaining a hardware monitoring software for 18 years, that covers all models from all manufacturers (Brocade, Cisco, Dell-EMC, HP, Huawei, HP, NetApp, Oracle, Pure, Synology, etc. I can't list them all). We've dealt with all sorts of hardware instrumentation standards, from specific SNMP MIBs to DMTF's WBEM, to IPMI, to SMI-S, to Redfish and SMASH, to proprietary REST APIs, you name it. See https://www.sentrysoftware.com/docs/hardware-connectors/23/platform-requirements.html for a full list of we got covered. 

We've rewritten our proprietary hardware monitoring software as an [OpenTelemetry Collector distribution](https://www.sentrysoftware.com/products/hardware-sentry-opentelemetry-collector.html) (available for free, not yet open source because this takes some time). We've taken some time to redefine the metrics that can be collected following OpenTelemetry semantic convention, and that's why we're proposing this.

With our experience, we are pretty confident that the proposed metrics cover 95% of the cases and are good enough to represent any system available out there. You can trust us on this 😎 (although we're open to discuss, obviously, and that's what this PR is for).

Note: The current version of our product doesn't strictly follow the proposed convention. We want to agree on something before we spend time updating our code and all dashboards that rely on these metrics.

bertysentry commented:
@reyang Thank you for your time and the review!

Let me reply to a few comments:

> Use standard units

Done!

> Consider/Explore putting metadata as Resource instead of Attributes

You are right and I would happily add that to the convention, but I don't think we can have a *Resource* belonging to another *Resource*. In the current state of the semantic convention, all metrics are supposed to be attached to a host *Resource* (with the `host.name` attribute, notably). If we wanted to attach the disk metrics to a disk *Resource*, we would need to duplicate all the attributes of the host *Resource* into the disk *Resource*. Is this something that you recommend? In which case, I'll update the doc accordingly (and our code as well).

> Reference to other standards (e.g. S.M.A.R.T.).

Done for S.M.A.R.T.

bertysentry commented:
@tigrannajaryan Could you please re-execute the workflow? All checks should pass now. Thank you!

bertysentry commented:
Thank you @tigrannajaryan for running the checks.

Unfortunately, **markdown-link-check** fails with this error:

```
FILE: ./specification/metrics/semantic_conventions/hardware-metrics.md
  [✖] https://schemas.dmtf.org/wbem/cim-html/2.49.0/CIM_Battery.html

  ERROR: 1 dead links found!

  23 links checked.
  [✖] https://schemas.dmtf.org/wbem/cim-html/2.49.0/CIM_Battery.html → Status: 0
make: *** [Makefile:31: markdown-link-check] Error 1
Error: Process completed with exit code 2.
```

But this check runs just fine in my environment **and** the offending link is valid! So I'm not sure what to do here. Re-run the checks, hoping for better luck? 🤷‍♂️

tigrannajaryan commented:
> But this check runs just fine in my environment **and** the offending link is valid! So I'm not sure what to do here. Re-run the checks, hoping for better luck? 🤷‍♂️

Network error maybe. I will re-run.

bertysentry commented:
@tigrannajaryan @pirgeo Thank you for your comments. I pushed an update to remove the non-standard unit (and moved it to the description). Please let me know if you are okay with this change.

bertysentry commented:
@tigrannajaryan @pirgeo Please let me know if you're okay with the latest changes so I can *resolve* the corresponding conversations. Also, who else should I ask for a review? Thank you! 😊

bertysentry commented:
@tigrannajaryan Thank you for your comment (and again, sorry I haven't seen it earlier, I'm not getting any notification, it's weird).

I vote for **option 1**. As you said, the document is still in **experimental** state, so breaking changes are expected (it regularly happens with the Otel SDKs). We can add a comment for all *Enum*-like metrics; "HIGHLY SUBJECT TO CHANGE".

**I personally commit to updating this semantic convention once the *Enum* metric #1711 issue is settled. **

tigrannajaryan commented:
@open-telemetry/specs-metrics-approvers we need you opinion on this PR, especially [on the question about enums](https://github.com/open-telemetry/opentelemetry-specification/pull/2518#issuecomment-1122464390).

bertysentry commented:
@pirgeo @tigrannajaryan I updated the **hardware-metrics** semantic convention to remove the offending *Enum* metrics, and replaced with proper *Attributes*.

Let me know of any remaining issue! Thank you again for the discussion, comments and suggestions!

pirgeo reviewed (approved):
Thanks for updating this. I see your points for Enums, and agree that they would be a better fit here. Until they are available, however, I believe this is a valid solution that works with the current state of the specification and is in line with the other semantic conventions. Thank you!

bertysentry commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+| `hw.host.energy`              | Total energy consumed by the physical host, in joules                                 | J     | Counter                                           | Int64      |                  |                            |
+| `hw.host.heating_margin`      | Distance in degrees Celsius to the closest temperature warning threshold              | Cel   | Gauge                                             | Double     |                  |                            |
+| `hw.host.power`               | Power consumed by the physical host in Watts                                          | W     | Gauge                                             | Double     |                  |                            |
+| `hw.host.status`              | Overall operational status: `1` (true) or `0` (false) for each of the possible states |       | Gauge                                             | Int        | `state`          | `ok`, `degraded`, `failed` |
```
@tigrannajaryan The [@open-telemetry/specs-metrics-approvers](https://github.com/orgs/open-telemetry/teams/specs-metrics-approvers) returns 404. Maybe this team doesn't exist?

tigrannajaryan replied:
The team exists. Probably 404 for you because you do not have the corresponding Github permissions to see it.

bogdandrutu replied:
I would personally postpone this. I think we need a discussion about how to represent enum metrics, and if we should have a separate instrument. Can we delay adding this?

bertysentry replied:
@bogdandrutu Please see issue #1711, which has been opened 1 year ago, and which is far from being solved, and will require even more time to be specified and implemented. Realistically, we won't get that feature before at least another year.

Instead of letting vendors (like my company) go with their own specific metrics for another year, I strongly encourage publishing this first version of the semantic convention for hardware metrics in the **experimental** state.

**I commit personally to updating this convention for hardware metrics once the *Enum* discussion is settled.**

bogdandrutu replied:
But is not about updating documentation that worries me, is about dealing with the breaking change and needing to built a translator, etc.

I think you have incentive to push for enums to be defined :) and you can work on that...

bertysentry replied:
Ha ha! But... but... but this is so blatantly manipulative! 😂

Reality is that @pirgeo and @dyladan convinced me that *Enums* as a specific type weren't such a great idea, given the sate of the timeseries market today. Chances to see timeseries vendors adopt a new *Enum* type are very slim.

That's why I updated the semantic convention for hardware metrics to use the widely adopted format of *StateSet*.

bertysentry replied:
@bogdandrutu Okay fair enough. **When approximately** will we get a status on *Enums*? #1711 has got 1 like (mine) in 1 year. People don't seem too excited about it.

bertysentry commented:
@tigrannajaryan I just rebased on **main** and fixed 3 broken links in **datamodel.md** (which are unrelated to this PR, but which were preventing the workflows). Could you please re-run the workflow? Thank you!

bogdandrutu commented on specification/metrics/datamodel.md:
```diff
@@ -377,10 +377,10 @@ same kind. <sup>[1](#footnotes)</sup>
-2. [Gauge](https://github.com/open-telemetry/opentelemetry-proto/blob/v0.10.x/opentelemetry/proto/metrics/v1/metrics.proto#L192)
-3. [Histogram](https://github.com/open-telemetry/opentelemetry-proto/blob/v0.10.x/opentelemetry/proto/metrics/v1/metrics.proto#L211)
-4. [Exponential Histogram](https://github.com/open-telemetry/opentelemetry-proto/blob/27a10cd70f63afdbddf460881969f9ad7ae4af5d/opentelemetry/proto/metrics/v1/metrics.proto#L239)
+1. [Sum](https://github.com/open-telemetry/opentelemetry-proto/blob/c5c8b28012583fda55b0cb16f73a820722171d49/opentelemetry/proto/metrics/v1/metrics.proto#L247)
+2. [Gauge](https://github.com/open-telemetry/opentelemetry-proto/blob/c5c8b28012583fda55b0cb16f73a820722171d49/opentelemetry/proto/metrics/v1/metrics.proto#L241)
+3. [Histogram](https://github.com/open-telemetry/opentelemetry-proto/blob/c5c8b28012583fda55b0cb16f73a820722171d49/opentelemetry/proto/metrics/v1/metrics.proto#L260)
+4. [Exponential Histogram](https://github.com/open-telemetry/opentelemetry-proto/blob/c5c8b28012583fda55b0cb16f73a820722171d49/opentelemetry/proto/metrics/v1/metrics.proto#L270)
```
Why this change? Consider to split in a separate PR.

bertysentry replied:
Without this change, the workflow fail because these links are no longer available, which will prevent merge of this PR. Please note that **main** branch is broken as well.

bertysentry replied:
I reverted this change and submitted a new PR. Please review #2552 and merge when possible, thanks!

bogdandrutu replied:
Done.

bogdandrutu commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+
+| Name                  | Description                                                                   | Units | Instrument Type ([*](README.md#instrument-types)) | Value Type | Attribute Key | Attribute Values           |
+| --------------------- | ----------------------------------------------------------------------------- | ----- | ------------------------------------------------- | ---------- | ------------- | -------------------------- |
+| `hw.enclosure.status` | Operational status: `1` (true) or `0` (false) for each of the possible states |       | Gauge                                             | Int        | `state`       | `ok`, `degraded`, `failed` |
+| `hw.enclosure.status` | Intrusion status: `1` (true) or `0` (false) for each of the possible states   |       | Gauge                                             | Int        | `intrusion`   | `open`                     |
```
Line 128 and 129 is the same metric?

bertysentry replied:
Good point! We thought of using 2 *Attributes* for the `hw.enclosure.status` metric: `state` and `intrusion`. But I will simplify this and use `state` *Attribute* only.

bogdandrutu commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+| `hw.gpu.errors`             | Number of errors encountered by the GPU                                       | {errors} | Counter                                           | Int64      | `type`        | `corrected`, `all`                              |
+| `hw.gpu.io`                 | Received and transmitted bytes by the GPU                                     | By       | Counter                                           | Int64      | `direction`   | `receive`, `transmit`                           |
+| `hw.gpu.memory_limit`       | Size of the GPU memory                                                        | By       | Gauge                                             | Int64      |               |                                                 |
+| `hw.gpu.memory_utilization` | Fraction of GPU memory used                                                   | 1        | Gauge                                             | Double     |               |                                                 |
```
is this utilization or usage? see as a reference for the system metrics https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/system-metrics.md#systemmemory---memory-metrics

bertysentry replied:
I confirm it's `utilization`. GPU instrumentation often return this value in percentage, and not the absolute value in bytes. But we can add `hw.gpu.memory_usage` (or `hw.gpu.memory.usage`) as well, just in case. These metrics are not mandatory anyway.

bogdandrutu commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+
+## Metric Instruments
+
+### `hw.host.` - Physical host metrics
```
What do we think between "host" vs "machine"?

bertysentry replied:
Personally, I prefer "host" because it's still an entity that runs an operating system and is accessible through the network, and therefore has a *hostname*. 

Also, these metrics must be attached to a [Host Resource](https://opentelemetry.io/docs/reference/specification/resource/semantic_conventions/host/).

So using `hw.host.*` sounds more consistent with the Host Resource and its `host.*` attributes.

Note: The most common occurrence of "machine" in computing is "Virtual Machine", meaning that the term "machine" is counter-intuitively more associated to virtual than physical 😂

bogdandrutu commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+| `hw.host.ambient_temperature` | Ambient (external) temperature of the physical host                                   | Cel   | Gauge                                             | Double     |                  |                            |
+| `hw.host.energy`              | Total energy consumed by the physical host, in joules                                 | J     | Counter                                           | Int64      |                  |                            |
+| `hw.host.heating_margin`      | Distance in degrees Celsius to the closest temperature warning threshold              | Cel   | Gauge                                             | Double     |                  |                            |
+| `hw.host.power`               | Power consumed by the physical host in Watts                                          | W     | Gauge                                             | Double     |                  |                            |
```
Is this current consumption or total? If total is not a counter?

bertysentry replied:
Power is electrical power, measured in Watts. This could be dropped as we also have `hw.host.energy` which is a counter of Joules. We agree that `hw.host.power = rate(hw.host.energy)`, and that's how power should be calculated. 

The problem is that most vendors don't expose an energy counter in their instrumentation, but a power meter in Watts.

bogdandrutu replied:
Then is "current power"? What is the correlation between that and time?

bertysentry replied:
`hw.host.power` is electrical power, which is always expressed in Watts. 

`hw.host.energy` is energy, which is always expressed in Joules. It's a counter.

`Power = ∆ Energy / ∆ Time` (Watts is in fact J/s).

`Energy = Sum( Power(t).dt )`

The best and recommended way to calculate the energy consumed by a system is to report this energy as a counter of joules. But unfortunately, most server manufacturers do not report this metric and are only able to report the instant power, in Watts. 

Ideally, developers of hardware monitoring solutions should expose both metrics, even if that means reconstructing the `hw.host.energy` counter (in joules) from the measured instant power (in Watts).

bertysentry replied:
Good question. There is no standard. Power can be reported as instant values with an amperemeter and a voltmeter as `P = UI`. Polling an instant value every few seconds and calculating an energy usage from these values is not mathematically accurate, and we should encourage a joule counter. But we need to adapt to what hardware vendors do, and many of them report the instant power only. So at least, for vendors who report the instant power value, they will use the proper metric name.

bogdandrutu commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+| Name                 | Description                                                                   | Units | Instrument Type ([*](README.md#instrument-types)) | Value Type | Attribute Key | Attribute Values           |
+| -------------------- | ----------------------------------------------------------------------------- | ----- | ------------------------------------------------- | ---------- | ------------- | -------------------------- |
+| `hw.fan.speed`       | Fan speed in revolutions per minute                                           | {rpm} | Gauge                                             | Int        |               |                            |
+| `hw.fan.speed_ratio` | Fan speed expressed as a fraction of its maximum speed                        | 1     | Gauge                                             | Double     |               |                            |
```
Is this needed if you have the `fan.speed`?

bertysentry replied:
Unfortunately, some fans are able to report their speed only as a percentage of their maximum speed (and that maximum speed may not even be exposed). To strictly follow Otel's general semantic convention, we should have had `hw.fan.speed_limit` and `hw.fan.speed_utilization`. But we can't really talk of *utilization* when it comes to speed. That's why we opted for `_ratio`. Let me know if that is acceptable.

bogdandrutu commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+| `hw.logical_disk.errors`      | Number of errors encountered on this logical disk                             | {errors} | Counter                                           | Int64      |               |                            |
+| `hw.logical_disk.limit`       | Size of the logical disk                                                      | By       | Gauge                                             | Int64      |               |                            |
+| `hw.logical_disk.status`      | Operational status: `1` (true) or `0` (false) for each of the possible states |          | Gauge                                             | Int        | `state`       | `ok`, `degraded`, `failed` |
+| `hw.logical_disk.usage`       | Logical disk space usage                                                      | By       | Gauge                                             | Int64      | `state`       | `used`, `free`             |
```
UpDownCounter

bertysentry replied:
I'm sorry, I read and read again the [UpDownCounter description](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#updowncounter) and I fail to see how it applies to disk or filesystem usage.

Disk usage is an absolute value and is not calculated as the sum of successive increments and decrements.

But I really must be missing something here, as I see **system-metrics.md** uses that for memory usage, paging usage, etc.

bogdandrutu replied:
> Disk usage is an absolute value and is not calculated as the sum of successive increments and decrements.

How is it calculated that "absolute" value? Is it not a Sum(+store - delete)? Same for memory is that not a sum (allocs - deallocs)?

bertysentry replied:
@bogdandrutu @tigrannajaryan I'm sorry to bother you on this, but what am I missing here?

bogdandrutu replied:
1. I did not say to use "delta" (we should report the cumulative), I said that it is a "sum" that we observe, so we should use UpDownCounter.
2. Read this https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#asynchronous-updowncounter, if does not help, I don't know how more to explain.

For your example you will report:
| Time | Reported/UpDownCounter Value |
|---|---|---|
| 8:00 | 1000 |
| 8:01 | 981 |
| 8:02 | 991 |
| 8:03 | 1005 |

bertysentry replied:
Okay now I understand. I was looking at *UpDownCounter*, but you were referring to *Asynchronous UpDownCounter*, which is slightly different (no deltas).

I'll update the convention accordingly. Thanks!

bogdandrutu commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+| --------------------------- | ----------------------------------------------------------------------------- | -------- | ------------------------------------------------- | ---------- | ------------- | ----------------------------------------------- |
+| `hw.gpu.errors`             | Number of errors encountered by the GPU                                       | {errors} | Counter                                           | Int64      | `type`        | `corrected`, `all`                              |
+| `hw.gpu.io`                 | Received and transmitted bytes by the GPU                                     | By       | Counter                                           | Int64      | `direction`   | `receive`, `transmit`                           |
+| `hw.gpu.memory_limit`       | Size of the GPU memory                                                        | By       | Gauge                                             | Int64      |               |                                                 |
+| `hw.gpu.memory_utilization` | Fraction of GPU memory used                                                   | 1        | Gauge                                             | Double     |               |                                                 |
```
Also I would do something like `hw.gpu.memory.limit` to be consistent with others.

bertysentry replied:
Good point. I was under the impression that we should limit the metric name to `<namespace>.<instrument>.<metric_name>`, but we can have more sub-items. I'll update all metrics accordingly.

bogdandrutu commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+| `hw.gpu.memory_limit`       | Size of the GPU memory                                                        | By       | Gauge                                             | Int64      |               |                                                 |
+| `hw.gpu.memory_utilization` | Fraction of GPU memory used                                                   | 1        | Gauge                                             | Double     |               |                                                 |
+| `hw.gpu.status`             | Operational status: `1` (true) or `0` (false) for each of the possible states |          | Gauge                                             | Int        | `state`       | `ok`, `degraded`, `failed`, `predicted_failure` |
+| `hw.gpu.time_utilization`   | Fraction of time spent in a specific task                                     | 1        | Gauge                                             | Double     | `task`        | `decoder`, `encoder`, `general`                 |
```
Maybe just `hw.gpu.utilization`? to be consistent with `system.cpu.utilization`?

bertysentry replied:
I was afraid of the confusion with `hw.gpu.memory_utilization`, but with `hw.gpu.memory.utilization` it just makes sense. Thanks!

bogdandrutu reviewed (commented):
Most of the comments apply to more than just the line where I added them, consider to fix all the places where they apply.

bertysentry commented:
Squashed commits and re-reverse change that was reversed during rebase 🤦‍♂️😅
Everything is clean now.

jmacd commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+
+| Attribute Key     | Description            | Example    |
+| ----------------- | ---------------------- | ---------- |
+| `sensor_location` | Location of the sensor | `CPU0_DIE` |
```
I wonder if hw.host.ambient_temperature can be classified as a temperature sensor, so as to avoid one special case?

jmacd reviewed (changes requested):
Nice work, @bertysentry.

bertysentry commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+| `hw.host.ambient_temperature` | Ambient (external) temperature of the physical host                                   | Cel   | Gauge                                             | Double     |                  |                            |
+| `hw.host.energy`              | Total energy consumed by the physical host, in joules                                 | J     | Counter                                           | Int64      |                  |                            |
+| `hw.host.heating_margin`      | Distance in degrees Celsius to the closest temperature warning threshold              | Cel   | Gauge                                             | Double     |                  |                            |
+| `hw.host.power`               | Power consumed by the physical host in Watts                                          | W     | UpDownCounter                                     | Double     |                  |                            |
```
I would be in favor of that, but most vendors equipped their hardware with a power sensor (reporting instant power in Watts) instead of a proper energy counter (in Joules). I just wanted to make sure this case is covered with a "standard" metric and that everybody uses the same metric name.

Also, instant power is an important core metric. For example, a datacenter administrator would like to know the sum of the maximum instant power of each system over time (more precisely: `sum_over_hosts(max_over_time(hw.host.power))`). Depending on the timeseries backend, this sort of calculation may not be possible with `rate(hw.host.energy)` if you can't specify the time steps to calculate the rate.

I won't fight hard for this, but my opinion is that a standardized `hw.host.power` will make administrator's life easier in some cases.

bertysentry replied:
Geez I can't get my head around *UpDownCounter* vs *Gauge* thing... 
@bogdandrutu @jmacd `hw.host.power` being the instantaneous power in Watts of a system, it can be summed up across a group of hosts. Should it be a *Gauge* or an *UpDownCounter*? (I really fail to see the difference it makes for either storage, querying or visualization)

jmacd replied:
> Should it be a Gauge or an UpDownCounter? (I really fail to see the difference it makes for either storage, querying or visualization)

OTel has a problem in explaining this--not your fault.

We came to this distinction originally through API mechanics. It would help explain our case if OTel supported a synchronous Gauge instrument. Then I could ask, to your question, which synchronous instrument would you choose? You would choose the Gauge instrument because the number is calculated, not counted. You would set the Gauge to the current power measurement. You would not use a Counter or UpDownCounter to count Watts going in (or out) of your system, because that quantity is derived from something else you could actually count.

This case, where an instantaneous quantity is has units that are `units/second` or `{things}/second`, we end up with something that can be meaningfully summed because (a) the underlying quantity has additive structure, and (b) the rate of time is equal across the hosts. When measurements are derived from time this way, the Gauge/UpDownCounter distinction breaks down, though the test above still applies.

bertysentry replied:
Thank you @jmacd! I think I have a better understanding of the difference. To decide the type of metric, we need to consider being as close as possible to the thing being observed, not intermediary instrumentation layers that may report the absolute value. I'll review all hardware metrics to make sure they follow this pattern.

bertysentry commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+| Name                          | Description                                                                           | Units | Instrument Type ([*](README.md#instrument-types)) | Value Type | Attribute Key(s) | Attribute Values           |
+| ----------------------------- | ------------------------------------------------------------------------------------- | ----- | ------------------------------------------------- | ---------- | ---------------- | -------------------------- |
+| `hw.host.ambient_temperature` | Ambient (external) temperature of the physical host                                   | Cel   | Gauge                                             | Double     |                  |                            |
+| `hw.host.energy`              | Total energy consumed by the physical host, in joules                                 | J     | Counter                                           | Int64      |                  |                            |
```
Joule is [international standard unit for energy](https://ucum.org/ucum.html#si), and yes, it is strictly defined as Watt-seconds. It's also recommended as a [standard metric in Prometheus](https://prometheus.io/docs/practices/naming/#base-units). It's also standard in Grafana visualizations and others, which easily convert Joules to the more common kWh (or MWh).

As per Otel general semantic convention, we need to stick to Joules. We would add a comment that 1 Wh = 3600 Joules if that helps.

bertysentry replied:
Energy counter sensors often report their values in Joules:

* [IPMI](https://openipmi.sourceforge.io/IPMI.pdf)
* [Redfish](http://redfish.dmtf.org/schemas/DSP2046_2020.1.html) (which also has kWh, but which I haven't seen in the wild, except for PDUs)

So in most cases, only 1 conversion will be necessary for visualization 😉

bertysentry commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+| ----------------------------- | ------------------------------------------------------------------------------------- | ----- | ------------------------------------------------- | ---------- | ---------------- | -------------------------- |
+| `hw.host.ambient_temperature` | Ambient (external) temperature of the physical host                                   | Cel   | Gauge                                             | Double     |                  |                            |
+| `hw.host.energy`              | Total energy consumed by the physical host, in joules                                 | J     | Counter                                           | Int64      |                  |                            |
+| `hw.host.heating_margin`      | Distance in degrees Celsius to the closest temperature warning threshold              | Cel   | Gauge                                             | Double     |                  |                            |
```
Good point, I agree the description is not clear. I'll update. 

This logic is commonly used in temperature sensors to report the "distance to failure" (Intel uses [Distance to TJMax](https://www.kernel.org/doc/Documentation/hwmon/coretemp) for example), 

We're applying this logic to an entire host in an effort to easily report whether a computer is needlessly cold, which means the ambient temperature could be safely increased by a few degrees, thus saving energy on air conditioning (1°C higher in ambient temperature means 5% saving on energy spent in cooling, and therefore in carbon emissions).

Example: All your systems have a `hw.host.heating_margin` above 10°C. This means the datacenter admin could raise the ambient temperature by 5°C safely and cut energy costs and carbon emissions by 25% on air conditioning, i.e. 10% of the overall datacenter.

The relation with `hw.host.ambient_temperature` is that `hw.host.ambient_temperature + hw.host.heating_margin` is the temperature at which point you will theoretically get an alert on at least 1 temperature sensor in the monitored system.

Basically we're encouraging exposing a metric that can be leveraged to save the planet! 🥰

bertysentry commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+| Name                 | Description                                                                   | Units    | Instrument Type ([*](README.md#instrument-types)) | Value Type | Attribute Key | Attribute Values                                |
+| -------------------- | ----------------------------------------------------------------------------- | -------- | ------------------------------------------------- | ---------- | ------------- | ----------------------------------------------- |
+| `hw.cpu.errors`      | Total number of errors encountered and corrected by the CPU                   | {errors} | Counter                                           | Int64      |               |                                                 |
+| `hw.cpu.power`       | CPU current power consumption in Watts                                        | W        | UpDownCounter                                     | Double     |               |                                                 |
```
Agreed, see my comment above.

bertysentry commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+| Name                 | Description                                                                   | Units | Instrument Type ([*](README.md#instrument-types)) | Value Type | Attribute Key | Attribute Values           |
+| -------------------- | ----------------------------------------------------------------------------- | ----- | ------------------------------------------------- | ---------- | ------------- | -------------------------- |
+| `hw.fan.speed`       | Fan speed in revolutions per minute                                           | {rpm} | Gauge                                             | Int        |               |                            |
+| `hw.fan.speed_ratio` | Fan speed expressed as a fraction of its maximum speed                        | 1     | Gauge                                             | Double     |               |                            |
```
Some fans only report their speed in percentage of their maximum speed. Most often the maximum speed of a fan is not known (not exposed in a proper way). We could specify a `hw.fan.speed_limit` but it's going to be very rarely used.

Following Otel's general semantic convention, we should have `hw.fan.speed_utilization` but this sounded weird. That's why I opted for `hw.fan.speed_ratio` (which is admittedly Prometheus' lingo).

Please note that server manufacturers often trigger an alert when `hw.fan.speed_ratio` goes too low, indicating reduced cooling capacity. When value is 100%, it's not considered "saturated", but simply spinning at nominal speed. That's why `limit` and `utilization` seemed a little off in this case.

Let me know if you want me to update with `limit` and `utilization`.

bertysentry commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+| `hw.network.io`                    | Network traffic in bytes                                                                                     | By        | Counter                                           | Int64      | `direction`   | `receive`, `transmit`       |
+| `hw.network.packets`               | Network traffic in packets (or frames)                                                                       | {packets} | Counter                                           | Int64      | `direction`   | `receive`, `transmit`       |
+| `hw.network.status`                | Operational status, regardless of the link status: `1` (true) or `0` (false) for each of the possible states |           | Gauge                                             | Int        | `state`       | `ok`, `degraded`, `failed`  |
+| `hw.network.up`                    | Link status: `1` (up) or `0` (down)                                                                          |           | Gauge                                             | Int        |               |                             |
```
Okay, I'll update accordingly (and same for all `status` metrics)

bertysentry commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,335 @@
+| Name                                 | Description                                                                                               | Units | Instrument Type ([*](README.md#instrument-types)) | Value Type | Attribute Key | Attribute Values           |
+| ------------------------------------ | --------------------------------------------------------------------------------------------------------- | ----- | ------------------------------------------------- | ---------- | ------------- | -------------------------- |
+| `hw.temperature.status`              | Whether the temperature is within normal range: `1` (true) or `0` (false) for each of the possible states |       | Gauge                                             | Int        | `state`       | `ok`, `degraded`, `failed` |
+| `hw.temperature.temperature_celsius` | Temperature in degrees Celsius                                                                            | Cel   | Gauge                                             | Double     |               |                            |
```
Well, [base temperature unit is Kelvin](https://ucum.org/ucum.html#baseunits). Americans use Fahrenheit. Too avoid confusion, I specified `celsius` as it is permitted by Otel's general semantic convention.

Let me know if you still think `.celsius` should be removed. Thanks!

bertysentry replied:
I totally agree we should use Kelvin for temperatures. I was just arguing that some users may say `hw.temperature.temperature` should be in Kelvin. Americans may think it will be in Fahrenheit. That's why I thought adding `_celsius` would help avoid confusion, as [this section of the general semantic convention for metrics](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md#units) seems to be allowing:

> Units may be included when it provides additional meaning to the metric name

Being European, I'm perfectly fine with dropping `_celsius`, I won't be confused personally. @bogdandrutu @tigrannajaryan  @pirgeo Are you okay with this?

bertysentry replied:
`_celsius` suffix has been removed.

bertysentry commented:
@jmacd Thank you for the thorough review and the comments. I pushed an update and I'm still waiting for a few answers in the above discussions.

bertysentry commented:
@tigrannajaryan @bogdandrutu Anything else you guys would like me to change? Do we still want to wait to settle on the *Enum* discussion before we merge this semantic convention (which remains in **experimental** status)?

bogdandrutu reviewed (dismissed):
@jmacd I am still not convinced that we should make the decision on how to represent "enum/state" like metrics in this PR, we should make that decision in the issue, document it in the api/sdk then use this representation/convention here. I suggest to remove it from the PR for the moment.

bertysentry commented:
@bogdandrutu Most of the metrics in this PR are *Enums*. Removing them is crippling the semantic convention for hardware metrics to the point it's useless. I suggest we put this PR on hold until you guys have settled on #1711.

In the meantime, my company will have to proceed with our own set of metrics (as originally proposed in this PR) in our next release. Fingers crossed you guys don't diverge too much from what we will have done on our end.

bertysentry commented:
Rebased on **main**

reyang commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,374 @@
+| `hw.physical_disk.endurance_utilization` | Endurance remaining for this SSD disk                                                       | 1        | Gauge                                             | Double     | `state`           | `remaining`                                     |
+| `hw.physical_disk.size`                  | Size of the disk                                                                            | By       | Gauge                                             | Int64      |                   |                                                 |
+| `hw.physical_disk.smart`                 | Value of the corresponding [S.M.A.R.T.](https://en.wikipedia.org/wiki/S.M.A.R.T.) attribute | 1        | Gauge                                             | Int        | `smart_attribute` | `Seek Error Rate`, `Spin Retry Count`, etc.     |
+| `hw.status`                | Operational status: `1` (true) or `0` (false) for each of the possible states               |          | UpDownCounter                                     | Int        | `state`           | `ok`, `degraded`, `failed`, `predicted_failure` |
```
```suggestion
| `hw.status`                              | Operational status: `1` (true) or `0` (false) for each of the possible states               |          | UpDownCounter                                     | Int        | `state`           | `ok`, `degraded`, `failed`, `predicted_failure` |
```

bertysentry commented:
Bot, this is not over! 😅

bertysentry commented:
@bogdandrutu Any update on the *StateSet* vs *Enum* discussion?

carlosalberto commented:
Ping @bogdandrutu ;)

reyang commented:
> @bogdandrutu Any update on the _StateSet_ vs _Enum_ discussion?

Discussed during the Jun 28, 2022 Specificatino SIG Meeting. In general folks feel that stateset is the right direction instead of enum. Given there are many pending work on metrics, it is unlikely that stateset support will be prioritized very soon.

The suggestion from the SIG meeting - this is really a good addition and we should try to unblock it. Consider the following approach:
1. Update/remove the "direction" for I/O operations, to better align with the general direction #2617.
2. Update the wording and call out that "these are the values for states, eventually the plan is to use StateSet and we should revisit / make the change later".

pirgeo commented:
I wasn't able to join yesterday's SIG, so just for clarification: the StateSet would be an attribute value on the wire? Or is it intended as a separate wire data type?

bertysentry commented:
@pirgeo You will need to ask @jmacd who seems to be clear on how *StateSet* may be implemented in the future. However, until this is specified, *StateSet* will be implemented as it is suggested now in this PR (and not as *Enum* as originally suggested and discussed earlier in this PR).

bertysentry commented:
@bogdandrutu Conclusion of specification SIG meeting was to implement *StateSet* through *Attributes* as discussed in this PR (until *StateSet* are properly specified and implemented, which will not happen soon).

Please approve (or at least remove the "Request Changes" flag). I would like to get this merged as nothing else remains to be discussed on this.

Thank you for your help!

bertysentry commented:
@bogdandrutu Sorry for insisting, I know everyone's bandwidth is limited (see above `hw.network.bandwidth.limit` ha! 😅). All suggestions have been implemented. There is no discussion item that remains open. Literally, you just need to hit the "Approve", and "Merge" buttons here. Thank you!

jmacd commented:
I overode @bogdandrutu's block, but there aren't enough approvals. @open-telemetry/specs-metrics-approvers

bertysentry commented:
Thank you @jmacd! Who are the members of @open-telemetry/specs-metrics-approvers?

arminru commented:
> Thank you @jmacd! Who are the members of @open-telemetry/specs-metrics-approvers?

@bertysentry You can find them here: https://github.com/open-telemetry/community/blob/main/community-members.md#specifications-and-proto

jack-berg commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+| `hw.gpu.memory.limit`       | Size of the GPU memory                                                        | By       | Gauge                                             | Int64      |                      |                                                 |
+| `hw.gpu.memory.utilization` | Fraction of GPU memory used                                                   | 1        | Gauge                                             | Double     |                      |                                                 |
+| `hw.gpu.memory.usage`       | GPU memory used                                                               | By       | UpDownCounter                                     | Int64      |                      |                                                 |
+| `hw.gpu.power`              | GPU instantaneous power consumption in Watts                                  | W        | Gauge                                             | Double     |                      |                                                 |
```
It feels strange that this doesn't have the `.usage` suffix. 

And shouldn't it be an UpDownCounter instead of a gauge since its useful to sum watts used by various components?

bertysentry replied:
`hw.gpu.power` is expressed in Watts. If we had `hw.gpu.power.limit`, then we could have a `usage` metric, but that is rarely the case (such metric is rarely available in the instrumentation layer).

jack-berg replied:
It does appear that `.usage` is limited to metrics where the `.limit` is known. Not sure why that is, because a metric like this really feels more self describing with the `.usage` suffix.

jack-berg commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+| Name                      | Description                                                                   | Units | Instrument Type ([*](README.md#instrument-types)) | Value Type | Attribute Key(s)           | Attribute Values                                      |
+| ------------------------- | ----------------------------------------------------------------------------- | ----- | ------------------------------------------------- | ---------- | -------------------------- | ----------------------------------------------------- |
+| `hw.battery.charge`       | Remaining fraction of battery charge                                          | 1     | Gauge                                             | Double     |                            |                                                       |
+| `hw.battery.charge.limit` | Lower limit of battery charge fraction to ensure proper operation             | 1     | Gauge                                             | Double     | `limit_type` (recommended) | `critical`, `throttled`, `degraded`                   |
```
Odd that limit is a lower limit in this context, but an upper limit probably everywhere else. Probably fine for now, but likely to change when #2557 is resolved.

bertysentry replied:
Well, it's natural to set a lower threshold for a battery charge. That's what all instrumentation layers out there do for batteries. I've never seen the battery charge expressed as a percentage of depletion and upper alert thresholds (like warning if 80% depleted).

jack-berg replied:
Right. If in the future we replace `.limit` with `.upper_limit` and `.lower_limit`, I'd expect this to change to `.lower_limit`.

jack-berg commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+| ----------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- | ----- | ------------------------------------------------- | ---------- | ---------------- | ---------------- |
+| `hw.host.ambient_temperature` | Ambient (external) temperature of the physical host                                                                                                   | Cel   | Gauge                                             | Double     |                  |                  |
+| `hw.host.energy`              | Total energy consumed by the physical host, in joules                                                                                                 | J     | Counter                                           | Int64      |                  |                  |
+| `hw.host.heating_margin`      | By how many degrees Celsius the temperature of the physical host can be increased, before reaching a warning threshold on one of the internal sensors | Cel   | Gauge                                             | Double     |                  |                  |
```
Is this the margin that `hw.host.ambient_temperature` can increase before crossing the threshold? If so, it would be more consistent to model it as a limit instead of a margin, i.e. `hw.host.ambient_temperature.limit`.

bertysentry replied:
No, it's not related to the ambient temperature. It's the minimal distance between each temperature sensor value and its corresponding warning threshold. It's literally the margin available for heating for the system. For individual sensors, it is sometimes expressed as "distance to warning", or "distance to failure", or variants of this.

jack-berg commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+|                             |                                                                               |          |                                                   |            | `hw.type` (required) | `gpu`                                           |
+| `hw.gpu.io.receive`         | Received bytes by the GPU                                                     | By       | Counter                                           | Int64      |                      |                                                 |
+| `hw.gpu.io.transmit`        | Transmitted bytes by the GPU                                                  | By       | Counter                                           | Int64      |                      |                                                 |
+| `hw.gpu.memory.limit`       | Size of the GPU memory                                                        | By       | Gauge                                             | Int64      |                      |                                                 |
```
Should this be a UpDownCounter since its useful to sum gpu memory limits across devices?

Also applies to:
- `hw.logical_disk.limit`
- `hw.memory.limit`
- `hw.network.bandwidth.limit`
- `hw.physical_disk.size`
- `hw.power_supply.limit`

bertysentry replied:
TBH each time someone mentions *UpDownCounter*, I get a new understanding of this metric type.

Does this mean that *Gauges* are to be limited to fractions and percentages?

@pirgeo @reyang @tigrannajaryan Please confirm that the metrics listed above must be specified as *UpDownCounters* (they can indeed be summed up across devices, as suggested by @jack-berg)

pirgeo replied:
The question I always ask myself when trying to distinguish between UpDownCounter and Gauge is: Does it make sense to sum the values:
* It makes sense to sum physical disk size, for example. If you sum the physical disk sizes from two devices, you still get a meaningful number (the total amount of disk space you have available)
* It makes no sense to add together two temperatures of hardware components: just because one component runs at 60°C and another one runs at 40°C it does not mean that your whole setup runs at 100°C. 

Gauges are therefore not limited to percentages (e.g. Temperatures are a good example)

There is a section in the supplementary guidelines about this: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/supplementary-guidelines.md#additive-property

bertysentry replied:
Oh, I get it now. This is not the place to discuss the matter of *UpDownCounter* but this can be summarized as:

* Use *UpDownCounter* for [extensive properties](https://en.wikipedia.org/wiki/Intensive_and_extensive_properties#Extensive_properties)
* Use *Gauge* for [intensive properties](https://en.wikipedia.org/wiki/Intensive_and_extensive_properties#Intensive_properties)

I'm not sure however why we do such a difference. The ability to calculate a sum on a metric is not that different from calculating the average.

`sum(temperature)` is meaningless, but `avg(temperature)` does mean something and may be of interest.

Anyway, I'll update the specification to use *UpDownCounter* instead of *Gauge* for these metrics.

Thank you!

bertysentry commented:
@jack-berg Thank you for the review! (sorry for the late reply, I was on vacation 😅)

bertysentry commented:
@jack-berg Thank you for the approval! As a member of the open-telemetry/instr-wg group, do you have a way to approve as this group? 

![image](https://user-images.githubusercontent.com/32521698/182250890-0640da9b-cc97-4f81-a15f-0f36e24856f4.png)

jack-berg commented:
Unfortunately I'm not a [memeber](https://github.com/orgs/open-telemetry/teams/instr-wg/members)!

bertysentry commented:
@jack-berg That's too bad! You're officially listed here though: https://github.com/open-telemetry/community/blob/main/community-members.md#specifications-and-proto

bertysentry commented:
@tigrannajaryan How can we fix this? @jack-berg approved the PR, but as he's not really a member of the instr-wg group, his approval doesn't unlock the merge of this PR. Thanks!

bertysentry commented on CHANGELOG.md:
```diff
@@ -185,6 +185,8 @@ release.
   ([#2507](https://github.com/open-telemetry/opentelemetry-specification/pull/2507))
 - Map SunOS to solaris for os.type resource attribute
   ([#2509](https://github.com/open-telemetry/opentelemetry-specification/pull/2509))
+- Add hardware metrics
```
I'm sure this is necessary, as we're already under the `### Semantic Conventions` heading, but it doesn't hurt to be precise 😉

bertysentry commented:
@tigrannajaryan PR ready to be merged, thank you all for your help, reviews and suggestions! 😊

tigrannajaryan commented:
@open-telemetry/specs-approvers This PR formally has the required number of approvals. Since the PR is non-trivial let's going to keep it open for 2 more days to give time to object and if we don't hear anything let's merge.

@bogdandrutu since you are now back please comment if you have any additional thoughts.

arminru commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+
+| Attribute Key | Description                                                                                                   | Example                             | Requirement Level |
+| ------------- | ------------------------------------------------------------------------------------------------------------- | ----------------------------------- | ----------------- |
+| `id`          | A unique identifier for the hardware component in the monitored host                                          | `win32battery_battery_testsysa33_1` | Required          |
```
Can you always expect to have a unique identifier for any hardware component and thus make it required?
It should also be specified how "unique" this is expected to be. If it's not globally unique but only _within_ the host identified by the `host` resource, it could be clarified like this:
```suggestion
| `id`          | An identifier for the hardware component, unique within the monitored host                                             | `win32battery_battery_testsysa33_1` | Required          |
```

bertysentry replied:
Good point!

arminru commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+| Name        | Description                                                                        | Units    | Instrument Type ([*](README.md#instrument-types)) | Value Type | Attribute Key(s)   | Attribute Values           |
+| ----------- | ---------------------------------------------------------------------------------- | -------- | ------------------------------------------------- | ---------- | ------------------ | -------------------------- |
+| `hw.energy` | Energy consumed by the component, in joules                                        | J        | Counter                                           | Int64      |                    |                            |
+| `hw.errors` | Errors encountered by the component                                                | {errors} | Counter                                           | Int64      | `type` (optional)  |                            |
```
```suggestion
| `hw.errors` | Number of errors encountered by the component                                      | {errors} | Counter                                           | Int64      | `type` (optional)  |                            |
```

arminru commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+| `hw.energy` | Energy consumed by the component, in joules                                        | J        | Counter                                           | Int64      |                    |                            |
+| `hw.errors` | Errors encountered by the component                                                | {errors} | Counter                                           | Int64      | `type` (optional)  |                            |
+| `hw.power`  | Instantaneous power consumed by the component, in Watts (`hw.energy` is preferred) | W        | Gauge                                             | Double     |                    |                            |
+| `hw.status` | Operational status: `1` (true) or `0` (false) for each of the possible states      |          | UpDownCounter                                     | Int        | `state` (required) | `ok`, `degraded`, `failed` |
```
Should one only report `1` for the state that applies or also `0` for the other two states?

bertysentry replied:
This question is still open until *StateSet* are specified and implemented in Otel. For now, it is recommended to push `0` for the other two states, at least once.

arminru replied:
I wonder what makes more sense for consumers but as long as the applicable states are reported with `1` it should be fine. I think we can leave it as-is, i.e., unspecified whether both should be reported or just `1`s.

arminru commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+| `hw.host.power`               | Instantaneous power consumed by the physical host in Watts (`hw.host.energy` is preferred)                                                            | W     | Gauge                                             | Double     |                  |                  |
+
+> **Note**
+> Host energy usage must be reported using the specific `hw.host.energy` and `hw.host.power` metrics instead of the generic `hw.energy` and `hw.power` described above, to avoid the risk of summing up overlapping values.
```
```suggestion
> Host energy usage MUST be reported using the specific `hw.host.energy` and `hw.host.power` metrics only, instead of the generic `hw.energy` and `hw.power` described above, to avoid the risk of summing up overlapping values.
```

arminru commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+| Name                          | Description                                                                                                                                           | Units | Instrument Type ([*](README.md#instrument-types)) | Value Type | Attribute Key(s) | Attribute Values |
+| ----------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- | ----- | ------------------------------------------------- | ---------- | ---------------- | ---------------- |
+| `hw.host.ambient_temperature` | Ambient (external) temperature of the physical host                                                                                                   | Cel   | Gauge                                             | Double     |                  |                  |
+| `hw.host.energy`              | Total energy consumed by the physical host, in joules                                                                                                 | J     | Counter                                           | Int64      |                  |                  |
```
```suggestion
| `hw.host.energy`              | Total energy consumed by the entire physical host, in joules                                                                                                 | J     | Counter                                           | Int64      |                  |                  |
```

arminru commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+| `hw.host.ambient_temperature` | Ambient (external) temperature of the physical host                                                                                                   | Cel   | Gauge                                             | Double     |                  |                  |
+| `hw.host.energy`              | Total energy consumed by the physical host, in joules                                                                                                 | J     | Counter                                           | Int64      |                  |                  |
+| `hw.host.heating_margin`      | By how many degrees Celsius the temperature of the physical host can be increased, before reaching a warning threshold on one of the internal sensors | Cel   | Gauge                                             | Double     |                  |                  |
+| `hw.host.power`               | Instantaneous power consumed by the physical host in Watts (`hw.host.energy` is preferred)                                                            | W     | Gauge                                             | Double     |                  |                  |
```
```suggestion
| `hw.host.power`               | Instantaneous power consumed by the entire physical host in Watts (`hw.host.energy` is preferred)                                                            | W     | Gauge                                             | Double     |                  |                  |
```

arminru commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+| `hw.battery.charge`       | Remaining fraction of battery charge                                          | 1     | Gauge                                             | Double     |                            |                                                       |
+| `hw.battery.charge.limit` | Lower limit of battery charge fraction to ensure proper operation             | 1     | Gauge                                             | Double     | `limit_type` (recommended) | `critical`, `throttled`, `degraded`                   |
+| `hw.battery.time_left`    | Time left before battery is completely charged or discharged                  | s     | Gauge                                             | Int        | `state` (recommended)      | `charging`, `discharging`                             |
+| `hw.status`               | Operational status: `1` (true) or `0` (false) for each of the possible states |       | UpDownCounter                                     | Int        | `state` (required)         | `ok`, `degraded`, `failed`, `charging`, `discharging` |
```
Note: Once we have support in our tooling to describe metrics in YAML definitions and create tables it might not be possible to define conflicting descriptions and attributes, so we might need to either add all attribute values to the "main" definition of `hw.status` above (Common section) or introduce separate `hw.battery.status` metrics and the like.

bertysentry replied:
Good point!

arminru reviewed (commented):
I left some non-blocking comments asking for clarification below.
Sorry for being late to the party - feel free to address them in a follow-up PR instead.

arminru commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+| Name              | Description                                                                   | Units    | Instrument Type ([*](README.md#instrument-types)) | Value Type | Attribute Key        | Attribute Values                                |
+| ----------------- | ----------------------------------------------------------------------------- | -------- | ------------------------------------------------- | ---------- | -------------------- | ----------------------------------------------- |
+| `hw.errors`       | Number of errors encountered on this memory module                            | {errors} | Counter                                           | Int64      | `hw.type` (required) | `memory`                                        |
+| `hw.memory.limit` | Size of the memory module                                                     | By       | UpDownCounter                                     | Int64      |                      |                                                 |
```
Would it make sense to report `usage` or `utilization` here as well?

bertysentry replied:
Nope, as memory utilization is strictly defined by the operating system.

arminru replied:
How does this differ from GPU or Disk usage/utilization?

bertysentry replied:
Good question!

`hw.memory.*` metrics represent memory modules, like you'd have a couple of 8-GB modules in your laptop. So `hw.memory.size` would be 8GB for 2 instances of the metric. Obviously, there is no concept of free or used capacity in a memory module, as the operating system allocates the entire system memory.

GPU memory management is performed at the lowest level by the GPU (and its driver). The operating system uses low-level calls to the GPU to allocate virtual memory on the GPU for each process. So there should be GPU memory utilization metrics both at the hardware level and at the operating system level. And the conclusion here is that there must be an `hw.gpu.memory.utilization` metric, as reported by the GPU itself, and therefore a corresponding `hw.gpu.memory.limit`.

`hw.physical_disk.*` metrics represent physical disks, and not filesystems. As such, there is no concept of free or used capacity (even for SSD, where it's just a counter of write operations on each block).

`hw.logical_disk.*` metrics represent logical storage entities to slice or group physical disks, and from which storage allocation is done, like RAID sets, LVM volumes, partitions or even filesystems (but that's a rare scenario for hardware instrumentation).

We really want the `.utilization`/`.limit` metrics combo to work together, to be "compliant" with the general semantic conventions.

arminru replied:
Thank you for the explanation!

arminru commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+| ---------------------------------------- | ------------------------------------------------------------------------------------------- | -------- | ------------------------------------------------- | ---------- | ------------------------------- | ----------------------------------------------- |
+| `hw.errors`                              | Number of errors encountered on this disk                                                   | {errors} | Counter                                           | Int64      | `hw.type` (required)            | `physical_disk`                                 |
+| `hw.physical_disk.endurance_utilization` | Endurance remaining for this SSD disk                                                       | 1        | Gauge                                             | Double     | `state` (required)              | `remaining`                                     |
+| `hw.physical_disk.size`                  | Size of the disk                                                                            | By       | UpDownCounter                                     | Int64      |                                 |                                                 |
```
`size` is inconsistent with the other attributes like `hw.logical_disk.limit` (_Size of the logical disk_), `hw.memory.limit` (_Size of the memory module_), `hw.gpu.memory.limit` (_Size of the GPU memory_).
Please use either of the two terms consistently.

bertysentry replied:
Yeah, I agree, but it's going to be difficult to make it 100% consistent with `.limit` **and** not sound weird: `hw.physical_disk.limit` sound weird to describe the size of a physical disk (and without corresponding `.utilization` metric).

`.limit` should be used only when there is a corresponding `.utilization` metric. In this case, it would be applicable to `hw.logical_disk.limit` and `hw.gpu.memory.limit`.

However, `hw.physical_disk.size` should remain *size*, and the same should go for `hw.memory.size` (because as I said in another comment, there is no utilization metric at the hardware level, it's purely an operating system metric).

**But** then we have `hw.memory.size` and `hw.gpu.memory.limit`, which is counterintuitive (and that's why we went with `hw.memory.limit`, for consistency).

To summarize:

| Current Metric | Corresponding `.utilization` metric? | New Suggestion |
|---|---|---|
| `hw.gpu.memory.limit` | Yes | `hw.gpu.memory.limit` |
| `hw.logical_disk.limit` | Yes | `hw.logical_disk.limit` |
| `hw.memory.limit` | No | `hw.memory.size`? |
| `hw.physical_disk.size` | No | `hw.physical_disk.size` |

arminru replied:
How about `.size` for all of them then? I think `hw.gpu.memory.size` and `hw.logical_disk.size` would not sound surprising. The utilization would then be a fraction of the overall physically available or logically assigned size.

bertysentry replied:
TBH I agree with you, but the general Otel semantic conventions command to use `.limit` when there is a `.utilization` metric. So we're just following that. Again, I agree that `.limit` is not the clearest word when you describe the size of a logical disk, but it's what has been done in many receivers.

arminru replied:
Fair! I think calling all `.limit` instead would follow the conventions as closely as possible and still be OK. I don't find it too weird personally to refer to the "size" of something as (physical) "limit".
The section you are referring to (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md#instrument-naming) only specifies what a `limit` is and that a `utilization` **can** be tracked based on that limit, but doesn't specify the `utilization` as a requirement for a `limit` metric, right?

bertysentry replied:
That's correct. Actually it says that `.usage` **should** *measure an amount used out of a known total (`.limit`)* and that *an instrument that measures the fraction of usage out of its limit should be called `.utilization`*. When to use `.limit` is not clearly defined, but `.usage` and `.utilization` **should** be used in relation to a `.limit` metric. You are correct that current convention doesn't stipulate that `.limit` should be used when there is a `.usage` or `.utilization` metric. But that should be clarified in the general semantic convention, so we don't need to discuss this at length in other semconv PRs...

So, are you saying I should rename `hw.physical_disk.size` to `hw.physical_disk.limit`? I'm not a big fan, but I won't fight against it 😉

Thanks!

arminru replied:
Given that the spec is not entirely clear on this and that we already have other metrics defined ending with `.size`, it should be fine. It just stuck out since both are used in the same PR here.
Since the other @open-telemetry/specs-metrics-approvers and @open-telemetry/instr-wg folks seemed to be fine with it as well, I think we can proceed and merge this.
Ideally, before moving the semantic conventions to a stable state, we would iterate over them once again and align any inconsistencies, where possible.

bertysentry replied:
Very cool! 😊

arminru commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+| ---------------------------------- | ------------------------------------------------------------------------------------------------------------ | --------- | ------------------------------------------------- | ---------- | -------------------- | --------------------------- |
+| `hw.network.bandwidth.limit`       | Link speed                                                                                                   | By        | UpDownCounter                                     | Int64      |                      |                             |
+| `hw.network.bandwidth.utilization` | Utilization of the network bandwidth as a fraction                                                           | 1         | Gauge                                             | Double     |                      |                             |
+| `hw.errors`                        | Number of errors encountered by the network adapter                                                          | {errors}  | Counter                                           | Int64      | `type` (recommended) | `all`, `zero_buffer_credit` |
+|                                    |                                                                                                              |           |                                                   |            | `hw.type` (required) | `network`                   |
```
In the other tables you list `hw.errors` as the first entry and we could keep this consistent here as well.

arminru commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+| `hw.network.bandwidth.utilization` | Utilization of the network bandwidth as a fraction                                                           | 1         | Gauge                                             | Double     |                      |                             |
+| `hw.errors`                        | Number of errors encountered by the network adapter                                                          | {errors}  | Counter                                           | Int64      | `type` (recommended) | `all`, `zero_buffer_credit` |
+|                                    |                                                                                                              |           |                                                   |            | `hw.type` (required) | `network`                   |
+| `hw.network.io.receive`            | Network traffic in bytes                                                                                     | By        | Counter                                           | Int64      |                      |                             |
+| `hw.network.io.transmit`           | Network traffic in bytes                                                                                     | By        | Counter                                           | Int64      |                      |                             |
+| `hw.network.packets.receive`       | Network traffic in packets (or frames)                                                                       | {packets} | Counter                                           | Int64      |                      |                             |
+| `hw.network.packets.transmit`      | Network traffic in packets (or frames)                                                                       | {packets} | Counter                                           | Int64      |                      |                             |
```
Please mentioned "received" and "transmitted" in the respective descriptions.

arminru commented on specification/metrics/semantic_conventions/hardware-metrics.md:
```diff
@@ -0,0 +1,383 @@
+|                                    |                                                                                                              |           |                                                   |            | `hw.type` (required) | `network`                   |
+| `hw.network.io.receive`            | Network traffic in bytes                                                                                     | By        | Counter                                           | Int64      |                      |                             |
+| `hw.network.io.transmit`           | Network traffic in bytes                                                                                     | By        | Counter                                           | Int64      |                      |                             |
+| `hw.network.packets.receive`       | Network traffic in packets (or frames)                                                                       | {packets} | Counter                                           | Int64      |                      |                             |
+| `hw.network.packets.transmit`      | Network traffic in packets (or frames)                                                                       | {packets} | Counter                                           | Int64      |                      |                             |
```
Should this say `{packets} or {frames}` in the Unit column or would you use {packets} in either case?

bertysentry replied:
Most backends won't be able to handle different units for the same metric, so we need to stick to `{packets}` (which can be safely considered a synonym of *frame*, or *datagram*, etc.)

arminru replied:
I agree. I think this would be the first convention with two different units for the same metric key and as you said, most backends will likely only support a single unit each.

bertysentry commented:
@arminru All suggestions have been taken into account, and branch rebased on *main*. Please review the few items open for discussion, thank you! 😊

bertysentry commented:
😭 <-- tears of joy
